### ISSUE TITLE
Extract PCM audio stream for waveform

### ISSUE THREAD

**Issue by beeradmoore** [linkedin/LiTr#119]
Heyya, 
I'm wanting to create a waveform based on a video selected by a user. We currently have this in iOS with the built in classes and it runs in about 0.33sec, however in our implementation on Android (non-LiTr) it is taking upwards of 11.41sec (it was 18.91sec before i added some threading into it 👀).

There appears to be a lack of info around how to do this efficiently on Android. I came across LiTr and was curious if its internal code is much better than mine and could be of more use.

Some background into my current Android code. Current code is in Xamarin.Android. This means I am using C# for my code so some class/method names may appear slightly different. For this usecase though the performance issues I am seeing should not be related to not using vanilla Java+Android. In my current code I am using a `MediaExtractor` + `MediaCodec` Decoder, selecting only the audio track and then in the `OnInputBufferAvailable` queuing that input buffer on the `MediaCodec` which then fires the `OnOutputBufferAvailable`. From here I can get the output buffer and loop over it to extract the shorts for me to parse and create a waveform.

At this point I am not even 100% that what I have is the raw audio data or if its just AAC data (eitherway it generates a the same waveform, so.. yay?). The iOS code works very differently. You tell it what output format you want, LinearPCM, 16bit, not as a float, one channel. And then it fires and I parse the result data in the exact same way.

My thought here is if I can use LiTr almost as the example in the readme is.
```
mediaTransformer.transform(requestId,
                           sourceVideoUri,
                           targetVideoFilePath,
                           targetVideoFormat,
                           targetAudioFormat,
                           videoTransformationListener,
                           transformationOptions);
```

Where `targetVideoFormat` is null because I just want to pass it through and `targetAudioFormat` is mono 16bit linear PCM. I didn't get very far with that because it's essentially transcoding the file. I'd still need to open it and parse the audio a then do my waveform. Neat if it's still faster than 11.41sec.

Looking around for PCM I only found [this](https://github.com/linkedin/LiTr/issues/4#issuecomment-562327327) comment that mentions decode audio frame into PCM. It's for a different purpose but same result. So I think what I need to do is to use the `TrackTransform` class. 

If I were to go this route would I extract the data with a `Renderer`, or would I some how capture this data after the `Encoder` has encoded it as PCM linear 16bit audio? Any heap or guidance with extracting PCM data would be greatly appreciated.

As a bonus I actually have setup a binding library over [here](https://github.com/beeradmoore/Xam.Plugin.LinkedIn.LiTr). After I've tidied it up and added some samples I'll release it as a nuget package so any other Xamarin.Android developer can use LiTr just like they would if they included it with gradle  👍

**Comment by izzytwosheds** [comment]
Interesting use case. So as I understand it, you are trying to extract PCM data from a video file? LiTr can help you, but with some tweaks:
 - you don't need to use `null` format for target video format. You can use `TrackTransform` API instead and ignore video track completely - just use a single `TrackTransform` with audio track index as a `sourceTrack` and zero as `targetTrack`, LiTr will "strip out" all other tracks. This is effectively the same as your logic that selects a single audio track in `MediaExtractor` I wonder if this would speed things up even more, since we would not be manipulating video frames at all...
 - output of a `Decoder` is a 16 bit linear PCM. (So yes, you have a "raw" data after decoding) Number of channels depends on a source audio, could be mono or stereo. If you need mono and source is stereo, you will have to mux two channels. LiTr doesn't currently support that. I can add support for audio filters, and we can work on adding a filter that muxes two channels.
 - No need to make `Encoder` encode into PCM, since its input is already PCM. We can introduce a "passthrough" encoder instead. This should be very straightforward.
 - In case you need to write into a file, Android's `MediaMuxer` doesn't support WAV format. We can write a simple `MediaTarget` implementation for that, should be fairly straightforward. 

In a nutshell, it really sounds like your code is doing effectively the same what LiTr is doing, just in a very specialized way. LiTr is meant to be more general purpose yet very customizable tool, since it is meant to serve a wide audience. Naturally, I would prefer to collaborate with you to add your functionality to LiTr so that everyone can use it. Let me know how you want to proceed.

**Comment by izzytwosheds** [comment]
Also, a huge thank you for Xamarin integration!

**Comment by izzytwosheds** [comment]
Also, which iOS APIs are you guys using? Converting entire audio track to PCM in under a second sounds unbelievable!

**Comment by beeradmoore** [comment]
Thanks for the response, sounds like I am on the right path at least! When I get time I'll try play around with  `TrackTransform` and `Decoder`. For my purposes I think I'll take the max of the front left and front right per sample for stereo (and 5.1 if someone happens to import that) as we only display our waveform as a single wave.

Regardless of the outcome of the audio waveform, we were also looking to downsample 4K to 1080p videos before the user uploads them. This is where I actually came across LiTr in the first place. If I can get audio waveform improvements that is just a bonus. 


> Also, a huge thank you for Xamarin integration!

No worries. It'll be better when I finish writing readme, add samples, and finish bindings so parameters are named and not p0, p1, p2, etc (although it appears most come through fine via binding the javadocs which is good).



> Also, which iOS APIs are you guys using? Converting entire audio track to PCM in under a second sounds unbelievable!

For iOS we are using Xamarin.iOS as our main framework, so all the underlying classes are just the standard iOS ones. Here is what a rough cut extracting PCM data looks like there. As a cross platform developer its always interesting to see how each platform handles specific tasks. In this case it appears the transcoding work I am doing is all built in for iOS, but on Android I am adding it myself.


```
// Load the asset, create a reader, get audio tracks.
var asset = AVAsset.FromUrl(fileUrl);
var reader = AVAssetReader.FromAsset(asset, out error);
var tracks = asset.TracksWithMediaType(AVMediaType.Audio);

if (tracks.Length == 0 || error != null)
{
    // No audio track found or failed to load.
    return;
}

// I don't recall if this is required if we are getting the entire audio
reader.TimeRange = new CMTimeRange()
{
    Start = CMTime.FromSeconds(0, asset.Duration.TimeScale),
    Duration = CMTime.FromSeconds(asset.Duration.Seconds, asset.Duration.TimeScale)
};

// Set the output to mono Linear PCM as an Int16
var outputSettings = new AudioSettings()
{
    Format = AudioFormatType.LinearPCM,
    LinearPcmBitDepth = 16,
    LinearPcmBigEndian = false,
    LinearPcmFloat = false,
    NumberChannels = 1,
    LinearPcmNonInterleaved = false,
};

var output = new AVAssetReaderTrackOutput(tracks[0], outputSettings)
{
    AlwaysCopiesSampleData = false,
};

reader.AddOutput(output);

var bytesPerInputSample = sizeof(short);

// Start parsing the file
reader.StartReading();
while (reader.Status == AVAssetReaderStatus.Reading)
{
    // Get the next sample buffer
    var trackOutput = reader.Outputs[0] as AVAssetReaderTrackOutput;
    var sampleBufferRef = trackOutput.CopyNextSampleBuffer();

    if (sampleBufferRef != null)
    {
        // Copy the data to a buffer to one we can read from
        var blockBufferRef = sampleBufferRef.GetDataBuffer();
        var bufferLength = blockBufferRef.DataLength;
        var mutableData = NSMutableData.FromLength((nint)bufferLength);
        blockBufferRef.CopyDataBytes(0, bufferLength, mutableData.MutableBytes);

        for (int i = 0; i < sampleBufferRef.NumSamples; i++)
        {
            // This is a short/Int16 represantation of an audio sample.
            var data = Marshal.ReadInt16(mutableData.MutableBytes, (int)(i * bytesPerInputSample));            
        }

        // Cleanup all
        sampleBufferRef.Invalidate();
        sampleBufferRef.Dispose();
        sampleBufferRef = null;
        mutableData.Dispose();
        mutableData = null;
    }
}
```

**Comment by izzytwosheds** [comment]
Update: writing PCM audio into WAV file is coming soon.

**Comment by beeradmoore** [comment]
@izzytwosheds an update on all of this.

The repo with full Xamarin.Android binding is up and available [here](https://github.com/beeradmoore/Xam.Plugin.LinkedIn.LiTr), and the nuget package for any Xamarin.Android developer to use is available [here](https://www.nuget.org/packages/Xam.Plugin.LinkedIn.LiTr). The nuget package for filters is its own seperate thing and it is [here](https://www.nuget.org/packages/Xam.Plugin.LinkedIn.LiTr.Filters).

When an update comes through to this repo all I need to do is find and replace 1.4.18 with 1.4.19 in my repo, re-run the build nuget packages script and it will fetch updated aars and build both nugets for me to go upload. At some point I'd like to go through the entire demo app and convert it to Xamarin.Android and then try keep it as updated as possible as its a very useful library.

In terms of how this helped us out. Our app generates `videoLength / 4` thumbnails and then choses closest to display. We could improve this by only generating what we need. The setup of `FrameExtractParameters` and `VideoFrameExtractor` actually lets us do that a lot more easier now. 

For my demo video of 92 seconds (23 small thumbnails) our generation time went from 5.04sec down to 4.58sec. ~9% improvement. Not huge, but I'll take it. This could be improved A LOT as in this current view we only actually display 3 images 😂

In terms of waveform generation we had HUGE improvements. We got it from 17.54sec down to 5.67sec. ~67% improvement. There may be more improvements to have here such as making the audio mono before we count samples, reducing bitrate/sample rate as that level of accuracy isn't needed for our waveforms. I am sure there are other parts that can be optimised but that is what I have off the top of my head.

Thanks for the effort on your end to help this all happen 👍

**Comment by izzytwosheds** [comment]
Wow! Thank you for keeping us updated. I as well will take any performance improvement any day, the bigger the better. 
Credit for `VideoFrameExtractor` goes to my colleague @ReallyVasiliy Let him know if you have ideas how to improve it further. I am very happy that new audio stack is working out for you. I put a lot of work into it to make it reliable and performant, Oboe library really helped out with that. 
Do you use WAV file or Decoder output directly? Just curious.
